### PR TITLE
DOC: add section about how to write good error messages

### PR DESCRIPTION
**Pull request by tacaswell**
## PR Summary

After leaving a comment to make error messages verbose I felt inspired to write it up for the docs...

Adding generated API docs for the private API might be a bit controversial, but I think on net it is the right thing to do so that we if we do remove them we can be sure that the docs do not still advocate for their use and saying "there are some `check_*` functions" felt too vague).

### CODE REVIEW

File: doc/devel/contributing.rst [hunk 1]
```diff
@@ -543,3 +543,76 @@ example code you can load it into a file handle with::
...
+An example of a **bad** error message ::
+
+  raise ValueError("invalid value")
+
+Which tells the user that we do not like the input, but no hints as to why!
+Because Matplotlib is the base for so many other libraries the user may not
+have directly called Matplotlib or even explicitly generated the invalid value.
+In cases like this it may be possible to get into a debugging and from reading
+the stack trace sort out right local variable to check is to start to sort out
+what went wrong.
```
Comment (ksunden) [inline doc/devel/contributing.rst:572]: not sure what is meant by "to get into a debugging", and not sure it's made better in context by saying "to get into a debugger", which was my first guess.

Comment (tacaswell) [reply doc/devel/contributing.rst:572]: did mean "debugger".

File: doc/devel/contributing.rst [hunk 2]
```diff
@@ -543,3 +543,76 @@ example code you can load it into a file handle with::
...
+
+This message tells the user what (local) variable Matplotlib checked was and
+what value was passed in which can help users rapidly find bugs in their own
+code.  For example if they expected to have passed in an integer but the error
+messages reports the string ``'aardvark'``, they may know where in their code
+that string appears and have an edge to start pulling on to debug the problem.
+While this does tell the user what variable and value Matplotlib rejected, it still
+does not tell them why.
+
+An example of a **best** error message ::
```
[the remark this answers is not in the document]
Comment (jklymak) [reply doc/devel/contributing.rst:586]: Agree with this - the first draft is pretty verbose....

File: doc/devel/contributing.rst [hunk 3]
```diff
@@ -543,3 +543,78 @@ example code you can load it into a file handle with::
...
+An example of a **bad** error message ::
+
+  raise ValueError("invalid value")
+
+Which tells the user that we do not like the input, but no hints as to why!
+Because Matplotlib is the base for so many other libraries the user may not
+have directly called Matplotlib or even explicitly generated the invalid value.
+In cases like this it may be possible to get into an interactive debugger and
+from reading the stack trace sort out right local variable to check is to start
+to sort out what went wrong.  However, without access to the running process the user
```
Comment (QuLogic) [inline doc/devel/contributing.rst:572]: After the `and`, this sentence kind of gets a bit muddled.

File: doc/devel/contributing.rst [hunk 4]
```diff
@@ -543,3 +543,78 @@ example code you can load it into a file handle with::
...
+
+Which tells the user what input in broken, what was passed in, and the reason
+why the value was invalid.   All of this together gives the user the best chance
+of being able to quickly diagnose and fix their problem by only reading the
+error message.
+
+Internal helpers
+~~~~~~~~~~~~~~~~
+
+Matplotlib has number of helper functions in the ``matplotlib._api`` module
```
Comment (QuLogic) [inline doc/devel/contributing.rst:600]:
```suggestion
Matplotlib has a number of helper functions in the ``matplotlib._api`` module
```

File: doc/devel/contributing.rst [hunk 5]
```diff
@@ -543,3 +543,78 @@ example code you can load it into a file handle with::
...
+named as ``check_*`` which provide nice error messages for very common checks
```
Comment (oscargus) [inline doc/devel/contributing.rst:601]:
```suggestion
named as ``check_*`` which provide nice error messages for common checks
```
very seems superfluous?

File: doc/devel/contributing.rst [hunk 6]
```diff
@@ -543,3 +543,78 @@ example code you can load it into a file handle with::
...
+
+
+An example of a **better** error message ::
+
+  raise ValueError(f"{value=!r} is an invalid value")
+
+This message tells the user what (local) variable Matplotlib checked was and
+what value was passed in which can help users rapidly find bugs in their own
+code.  For example if they expected to have passed in an integer but the error
+messages reports the string ``'aardvark'``, they may know where in their code
```
Comment (oscargus) [inline doc/devel/contributing.rst:583]:
```suggestion
message reports the string ``'aardvark'``, they may know where in their code
```

**Comment by oscargus** [comment]
I'm missing info about dropping punctuation characters and if an error message should be in Sentence case or nor.

**Comment by anntzer** [comment]
Highly nitpicky, but my personal(?) view is to do what cpython does:
```
>>> object()()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'object' object is not callable
>>> object().bar
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'object' object has no attribute 'bar'
>>> getattr(object(), 42)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: attribute name must be string, not 'int'
```
i.e. no leading capital and no trailing punctuation.
However, if the error message spans multiple sentences, then all sentences should be capitalized and punctuated correctly.  I try to avoid such cases, e.g. I would reword the error message of check_shape (currently `'arg' must be 2D with shape (M, 2). Your input has shape (3, 3).`) to `'arg' must be 2D with shape (M, 2); your input has shape (3, 3)`

**Comment by anntzer** [comment]
I agree with others that this seems pretty verbose (though a good idea to document); a first try at editing (while keeping most of your wording) gives me

```rst
Error messages should err on the side on being verbose, descriptive, and
specific about what went wrong.  They should be informative enough so that a
typical user (not an expert of our API) can understand, debug, and fix their
problem based solely on the message, without the need to consult the online
documentation.

For example, input validation errors should include, where possible,
information about which input is invalid, what value was passed in, and why the
value is invalid, e.g. ::

    raise ValueError(f"{value=!r} is invalid, value must be a positive integer")

This message helps the user to quickly diagnose and fix their problem.

Remember that many other libraries are implemented on top of Matplotlib, and
therefore ``value`` may not even have been directly passed by the user, but
rather generated in some intermediate layer.  In such cases, this kind of
highly explicit messages can be particularly useful, compared to shorter
messages such as ::

    raise ValueError("invalid value")
    # or
    raise ValueError(f"{value=!r} is invalid, value must be a positive integer")
```

**Comment by tacaswell** [comment]
I took @anntzer 's wording whole sale and applied the other changes.

File: doc/devel/contributing.rst [hunk 7]
```diff
@@ -549,3 +549,57 @@ example code you can load it into a file handle with::
...
     import matplotlib.cbook as cbook
     fh = cbook.get_sample_data('mydata.dat')
+
+.. _verbose-error-messages:
+
+Error Messages
+--------------
+
+
+Error messages should err on the side of being verbose, descriptive, and
```
Comment (QuLogic) [inline doc/devel/contributing.rst:559]:
```suggestion
--------------

Error messages should err on the side of being verbose, descriptive, and
```

File: doc/devel/contributing.rst [hunk 8]
```diff
@@ -549,3 +549,57 @@ example code you can load it into a file handle with::
...
+
+.. autosummary::
+   :toctree: _as_gen
+   :template: autosummary.rst
+   :nosignatures:
+
+   _api.check_isinstance
+   _api.check_in_list
+   _api.check_shape
+   _api.check_getitem
```
Comment (QuLogic) [inline doc/devel/contributing.rst:605]: These appear to already have been documented somewhere since the build is failing.

**Comment by story645** [comment]
Can you leave licensing as the last entry since it's not really like the others?

File: doc/devel/coding_guide.rst [hunk 9]
```diff
@@ -318,3 +318,58 @@ and running the same script will display
...
+specific about what went wrong.  They should be informative enough so that a
+typical user (not an expert of our API) can understand, debug, and fix their
+problem based solely on the message, without the need to consult the online
+documentation.
+
+For example, input validation errors should include, where possible,
+information about which input is invalid, what value was passed in, and why the
+value is invalid, e.g. ::
+
+    raise ValueError(f"{value=!r} is invalid, value must be a positive integer")
```
Comment (timhoffm) [inline doc/devel/coding_guide.rst:339]: I suggest to include the parameter name. It's not always needed, but good for educational purposes. For example

https://github.com/matplotlib/matplotlib/blob/6e95994dec16146a68ebcf90a61889be9df6f744/lib/matplotlib/text.py#L1263-L1264

Edit, oh maybe you mean *value* as the parameter name? In that case I would use a more targeted name for the example.

We may also consider adopting a policy to enquote parameter names in single ticks. While there's no universal convention, this is the closest I have found for a pattern, see e.g.

- CPython https://github.com/search?q=repo%3Amatplotlib%2Fmatplotlib+%2FValueError%5C%28%22.*%27%5Cw%2B%27%2F&type=code
- matplotlib https://github.com/search?q=repo%3Amatplotlib%2Fmatplotlib+%2FValueError%5C%28%22.*%27%5Cw%2B%27%2F&type=code

File: doc/devel/coding_guide.rst [hunk 10]
```diff
@@ -318,3 +318,58 @@ and running the same script will display
...
+
+This message helps the user to quickly diagnose and fix their problem.
+
+Remember that many other libraries are implemented on top of Matplotlib, and
+therefore ``value`` may not even have been directly passed by the user, but
+rather generated in some intermediate layer.  In such cases, this kind of
+highly explicit messages can be particularly useful, compared to shorter
+messages such as ::
+
+    raise ValueError("invalid value")
```
Comment (timhoffm) [inline doc/devel/coding_guide.rst:349]: Maybe inline to de-emphasize?

```suggestion
messages such as ``raise ValueError("invalid value")``.
```

When skimming the section, you primarily notice the heading and the two code blocks, giving the impression that both are relevant. There's no indication that one is a negative example. 

image

File: doc/devel/coding_guide.rst [hunk 11]
```diff
@@ -318,3 +318,58 @@ and running the same script will display
...
+
+.. autosummary::
+   :toctree: _as_gen
+   :template: autosummary.rst
+   :nosignatures:
+
+   _api.check_isinstance
+   _api.check_in_list
+   _api.check_shape
+   _api.check_getitem
```
Comment (timhoffm) [inline doc/devel/coding_guide.rst:375]: This generates API documentation resulting in duplication
- https://output.circle-artifacts.com/output/job/99a2d267-7d06-42e8-9318-237d42c8dd16/artifacts/0/doc/build/html/devel/_as_gen/matplotlib._api.check_isinstance.html#matplotlib._api.check_isinstance
- https://output.circle-artifacts.com/output/job/99a2d267-7d06-42e8-9318-237d42c8dd16/artifacts/0/doc/build/html/api/_api_api.html#matplotlib._api.check_isinstance

The best would be if one could tell autosummary not to create new API doecumentation and just link to the existing ones, but I think that is not possible (correct me if I'm wrong).

As a workaround, I'd create a manual table or bullet list (with or without copied/hard-coded description).

An alternative with a little more work would be to structure https://matplotlib.org/stable/api/_api_api.html into topic sections and and link to the section on validation/error handling.

File: doc/devel/coding_guide.rst [hunk 12]
```diff
@@ -318,3 +318,58 @@ and running the same script will display
...
+highly explicit messages can be particularly useful, compared to shorter
+messages such as ::
+
+    raise ValueError("invalid value")
+
+
+Internal helpers
+~~~~~~~~~~~~~~~~
+
+Matplotlib has a number of helper functions in the ``matplotlib._api`` module
```
Comment (timhoffm) [inline doc/devel/coding_guide.rst:355]: link
```suggestion
Matplotlib has a number of helper functions in the `matplotlib._api` module
```

File: doc/devel/coding_guide.rst [hunk 13]
```diff
@@ -318,3 +318,58 @@ and running the same script will display
...
+named as ``check_*`` which provide nice error messages for common checks
+in Matplotlib.  Please use these when working on the Matplotlib code base, it
+both saves a bunch of duplicate code and ensures our error messages are
+consistent across the library.  Use them in your own code at your own risk, we
+consider them private API and reserve the right change them with no notice on
+any release.
```
Comment (timhoffm) [inline doc/devel/coding_guide.rst:361]:
```suggestion
consistent across the library.
```
I would leave this out here. It's in an underscored module and the module docstring also explicitly states this.

**Comment by melissawm** [comment]
Hi folks - can we bump this? I think it would be a nice addition. Happy to help push it forward in any way needed. Thanks!
